src/main/groovy/com/ullink/gradle/sonar/SonarScannerMsbuildPlugin.groovy

```diff
@@ -0,0 +1,94 @@
+    static final MANDATORY_ARGS = [n: 'sonar.projectName', v: 'sonar.projectVersion']
+
+    @Override
+    void apply(Project project) {
+        def scannerExtension = project.extensions.create('sonarqube', SonarScannerMsbuildExtension)
+        project.afterEvaluate {
+            def msbuild = project.tasks[scannerExtension.msbuildTaskName ?: DEFAULT_MSBUILD_TASK_NAME]
+            if (msbuild == null) {
+                project.logger.warn('Not setting up sonar-msbuild-scanner since no msbuild task was found. ' +
+                    "Please specify the name of the msbuild task using the 'msbuildTaskName' method.")
+            } else {
+                def scannerVersion = scannerExtension.scannerVersion ?: DEFAULT_SCANNER_VERSION
+                def scannerZipName = "sonar-scanner-msbuild-${scannerVersion}-net46.zip"
+                def scannerDir = new File(CACHE_DIR, scannerVersion)
+                def sonarScanner = new File(scannerDir, 'SonarScanner.MSBuild.exe')
+
+                // Since some properties may not be evaluated at this moment, let's initialize the arguments
+                // right before the init task's execution
+                def configureArgsTask = project.task('configureSonarProperties') {
+                    doLast {
+                        project.tasks.initSonarScanner.commandLine += buildArgs(project, scannerExtension.properties.properties)
+                    }
+                }
+                def initSonarScannerTask = project.task('initSonarScanner', type: Exec, dependsOn: configureArgsTask) {
+                    commandLine = [sonarScanner, 'begin']
+                }
+
+                // keep the 'sonarqube' name for compatibility
+                def sonarScannerTask = project.task('sonarqube', type: Exec) {
+                    commandLine = [sonarScanner, 'end']
+                }
+                project.task('sonar', dependsOn: sonarScannerTask) // alias for sonarqube
+
+                if (!sonarScanner.exists()) {
+                    CACHE_DIR.mkdirs()
+                    def scannerDownloadTask = project.task('downloadSonarScannerMsbuild', type: Download) {
+                        src "https://github.com/SonarSource/sonar-scanner-msbuild/releases/download/${scannerVersion}/${scannerZipName}"
+                        dest File.createTempDir()
+                    }
+                    def scannerCacheDir = new File(CACHE_DIR, scannerVersion)
+                    scannerCacheDir.mkdir()
+                    def unzipTask = project.task('unzipSonarScannerMsbuild', type: Copy, dependsOn: scannerDownloadTask) {
+                        from project.zipTree(new File(scannerDownloadTask.dest, scannerZipName))
+                        into scannerCacheDir
+                    }
+                    initSonarScannerTask.dependsOn unzipTask
+                }
+                msbuild.dependsOn initSonarScannerTask
+                msbuild.finalizedBy sonarScannerTask
```

Comment:
Looks like the design here is different from **sonar-scanner-gradle**
In the standard plugin, you don't get any sonarqube executed unless explicitly calling/depending on the sonarqube task

Is that on purpose?

Reply:
Actually it is. Sonar-scanner-msbuild needs to call `begin` before msbuild starts to hook into the MSBuild pipeline. This is a major difference compared to the original plugin which is independent of the build tasks (msbuild, compileJava, etc).
@muryoh do you know any way to add these two task dependencies only if the `sonarqube` task is actually called?
If no, I should at least document this behavior.

Reply:
One alternative would be to completely remove the dependencies on `msbuild` and make the plugin user's responsibility to add them. WDYT?

Reply:
I tried this approach: https://gist.github.com/muryoh/1e70d7ff5625fb8988fc58932b5a80b9
Note that I used the configuration avoidance API but it was just to try it out - not using it would be OK too
Basically the ideais to always add the task dependencies
initSonarScannerTask -> msbuildTask -> sonarquebeTask

Then in the build graph disable initSonarScannerTask if sonarqube is not in the graph (ie to be executed)

Reply:
Thanks @muryoh ! `whenReady` seems like a good way to achieve this. It still seems like a hack but it is a needed one :)
Actually I changed your solution a bit to get rid to the `afterEvaluate` block and the `msbuild` task dependency.